doc/services/pm/device.rst

```diff
@@ -10,37 +10,34 @@ setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
 selected, device drivers implementing power management will be able to take
 advantage of the device power management subsystem.
 
-Zephyr supports two types of device power management:
+Zephyr supports two methods of device power management:
 
  - :ref:`Device Runtime Power Management <pm-device-runtime-pm>`
- - :ref:`System Power Management <pm-device-system-pm>`
+ - :ref:`System Managed Device Power Management <pm-device-system-pm>`
 
 .. _pm-device-runtime-pm:
 
 Device Runtime Power Management
 *******************************
 
-In this method, the application or any component that deals with devices directly
-and has the best knowledge of their use, performs the device power management. This
-saves power if some devices that are not in use can be turned off or put
-in power saving mode. This method allows saving power even when the CPU is
-active. The components that use the devices need to be power aware and should
-be able to make decisions related to managing device power.
+In this method, it is assumed that devices have the best knowledge of their use
```

Comment:
I strongly disagree with saying that users of devices do not need to be power aware, and the statement that power management should be handled by the device driver.

Drivers themselves don't have any knowledge of future usage patterns, so they can't do anything smarter than "turn on for this sample/action then turn off again". Obviously this is not great if the action is an I2C bus read, and the user is planning to run hundreds of them.

The user of the device is in the best position to say what state the device should be in at a given time, not the device itself.

Reply:
Ok, changed the text to reflect that.

Reply:
The text is still inverting the current meaning, from `In this method, the application or any component that deals with devices directly and has the best knowledge of their use` to `In this method, it is assumed that devices have the best knowledge of their use`. Devices don't know anything about their use, thats application knowledge.

Reply:
ok, simplified further, note there are more details on runtime pm in the reference section, this is a just quick intro.

Reply:
I appreciate that, but the meaning is still being inverted.
From "users of the device should handle PM" to "devices should handle their own PM".
Since the docs tend to be the source of truth of how things are intended to be used, I don't think this should be changed without cause.

Reply:
I am a bit confused. Users of the device do not handle PM, the PM operations are all within the device driver and implemented in the driver, so the statement above is correct. Are we talking past eachother here?

Reply:
I think we are, as that has never been how I have used PM. And the number of drivers using `PM_DEVICE_DT_INST_DEFINE` without corresponding usage of `pm_device_runtime_get` on themselves would seem to back me up. Out of 72 drivers I think I saw 4 using the pattern you describe, including at least one that was obviously wrong (nrf_qspi_nor)

Reply:
Which is not how it has historically been documented/used, and is not how it is used in the majority of in-tree drivers:
https://docs.zephyrproject.org/2.7.5/reference/power_management/index.html#runtime-device-power-management
Reference counting PM usage (the point of runtime PM) seems redundant if it is only used internally.

Reply:
agree we need more drivers using runtime pm and fixing the documentation is the first step. Now you make wonder how you do driver PM yourself if you are not using runtime pm?

Reply:
> agree we need more drivers using runtime pm and fixing the documentation is the first step

I am in agreement with that.

> Now you make wonder how you do driver PM yourself if you are not using runtime pm?

I am using runtime PM across all of my applications, on all but the simplest sensors, in the manner that I describe.
* Drivers wrap N transactions on SPI/I2C buses with `pm_device_runtime_get/put()`
* "Simple Sensor" (Temperature, etc) threads wake the sensor with `pm_device_runtime_get`, run N samples for averaging,  release with `pm_device_runtime_put`.
* "Complex Sensor" (IMU, etc) threads wake the sensor with `pm_device_runtime_get`, leave it in active mode for N seconds, then power it down with `pm_device_runtime_put`.
* Data logging batches data in RAM, then flushes across N writes to Flash/UART, all wrapped in one sequence of `get/put`
* etc

I care about this because I use it everywhere, and the recommendation being made doesn't fit with my experience as a user. I have been making Zephyr low power on every version since 2.3, well before it was properly supported in-tree.

Reply:
> OTOH, in some cases like entropy generation which works asynchronously with the application (to maintain an entropy buffer), implementation within the driver is the only solution.

Sure, there are edge cases in any setup, I agree performing the PM in that driver is appropriate for the situation. But what is the "common" use-case, a driver hidden behind a library called occasionally (entropy behind sys_random), or a sensor that you want to read from, or a bus you want to transmit over? My issue is with the default recommendation.

Reply:
The other problem I have is that it is not necessarily possible for the driver to determine when it should be calling `get/put` on itself. Many devices lose internal state when put into sleep modes.

How does a function like  `sensor_attr_set` operate in a world where the device is expected to run PM on itself?
* If it does both PM operations, any configuration can be lost immediately. 
* If it doesn't, all configuration needs to be stored in RAM somewhere until it is called. When does that happen? On `sensor_fetch`? Are we reconfiguring the entire chip each time we fetch a sample?
* What about if it only calls `get` to preserve the state. Who is responsible for calling `put` (The same number of times as we ran a configure function)? And when?

All trivially solved with the requirement that users of the device are explicit about when it should be active, very difficult to resolve otherwise.

Reply:
The idea with device runtime pm is the device saves energy when it is not being used, this works well for devices that are called from time to time, obviously devices have some limited knowledge of when they will be used so, upper layers in the stack can help with it, if they know that they will need the device and it can't be suspended they should `get()` the device.

Note that the driver has to be aware of its capabilities, as you mention the sensor example, if a sensor loses context when suspending itself, it clearly cannot suspend  itself ...

I think we need to document that this method goal is that devices are responsible to save energy as much as they can / are possible and that components that deal with them can interfere holding or releasing references to the device pm.

The way I see, a  generic device will call `get(self)` when they are requested and `put(self)` after perform the operation. If the component that is calling the device knows that the device does not need to suspend in each operation, this component should call `get(dev)` while using the device and releasing `put(dev)` when it is no longer needed.

Corner cases will always exist, if a device lose context and it can compromise its usability, it probably should not perform device runtime operations in itself and rely on upper components to do it.

Reply:
> Note that the driver has to be aware of its capabilities, as you mention the sensor example, if a sensor loses context when suspending itself, it clearly cannot suspend itself ...

Now take the example a step further. You have two boards, identical except for a different temperature sensor. One loses state on power down, the other doesn't.

The whole purpose of the sensor API is to provide a common interface to broadly similar devices, so this should be no problem. But if you want code you write to work across both boards, you have no choice but to work in the way I suggest. 

If the drivers are following the model of "call `get(self)` only if you can", and the user does not explicitly call PM functions, using the sensor which does lose state will always fail. Whereas if the user always calls PM explicitly, it works for both, even if the driver that preserves state is not calling `get(self)`.

For a robust system, you don't have the option of choosing on a per-implementation basis what the behaviour should be, it needs to be specified at the API level. My argument would be that any API with linked call structures must be driven by external users, while simple API's can be done internally, with the obvious option of users running higher level PM operations to optimise use cases.

Some examples:
* Sensor: Complex API, many interrelated function calls, externally driven
* Entropy: Simple API, independent function calls, internally driven
* Flash: Simple API, independent function calls, internally driven
* SPI/I2C: Simple API, independent function calls, internally driven
* UART: Complex API, interrelated calls/events, externally driven

While most API's probably fall under the simple case, by volume of drivers that people (end-users especially) write, I would argue that complex APIs (Sensor) dominate, and any generic recommendations in the docs should be tailored towards that.

Reply:
I like this, I just ran into something similar with DMA.

I believe DMA should be externally driven, e.g. the thing using DMA should call get/put as it knows when the transfers are being done presumably. There may be cases where polling is used in place of interrupts and in such cases, truly only the code polling the DMA (get_status()) would know!

Reply:
I mostly agree, with the usual amount driver specific cases. See https://github.com/zephyrproject-rtos/zephyr/pull/63158 for instance (UART: externally driven, but additionally requires implementation specific internal drive).

Reply:
I am in agreement here. Devices should do the best they can and handle their peculiarities like @erwango said, though subsystems and more elaborated APIs have to be responsible for the power management as well, to avoid inconsistencies between different drivers.

What I am saying is that the general thumb rule is something like devices should do what they can to save energy (and be usable) and higher layers should also do it. DMA is one example indeed, the API calls may not be balanced so the device can easily get into a point where the pm reference count is wrong, so the user of this driver has to be one in control.

It is hard to define a line about who is responsible for pm, the device or the subsystem. So the idea is that both should be doing  this.